Replay parse actions before loading data

emily-howell opened the pull request.

emily-howell committed:
Initial function for loading in the actions file before reading data

emily-howell commented:
I'm not 100% sure that this is the direction that #1317 intended. I know there is the "CheckForSettingsFile" function in ascent.cpp, I was hesitant to expose that function but alternatively could and just use that.

A few questions on this code:
1) Are there additional steps that should be added if ASCENT_REPLAY_MPI is defined. I am a little rusty on my MPI so some feedback there would be appreciated.
2) There aren't any existing tests for the replay functionality as far as I can see. Should I add something for this somewhere?

emily-howell committed:
Adding in MPI handling to actions file loading

cyrush commented on src/utilities/replay/replay.cpp:
```diff
@@ -156,6 +156,66 @@ void trim(std::string &s)
+  
+  bool has_file = false;
+
+  if(conduit::utils::is_file(file_name)){
```
We should only check if the file exists on rank 0.

cyrush commented on src/utilities/replay/replay.cpp:
```diff
@@ -156,6 +156,66 @@ void trim(std::string &s)
+    ASCENT_ERROR("Failed to load actions file: " << file_name
+                  << "\n" << file_load_error_msg);
+  }
+
```
If successful loading the file on rank 0, we need to broadcast the actions to all other ranks.
Conduit's relay MPI allows us to broadcast a node.

cyrush commented on src/utilities/replay/replay.cpp:
```diff
@@ -156,6 +156,66 @@ void trim(std::string &s)
 }
 
+conduit::Node load_actions(const std::string &file_name, int mpi_comm_id){
+  conduit::Node actions;
```
We prefer passing in a ref to the node and having node filled out instead of returning a node.

This is b/c returning a node involves more copies.

cyrush commented on src/utilities/replay/replay.cpp:
```diff
@@ -197,16 +257,16 @@ main(int argc, char *argv[])
+  // Populate actions with actions file
   //
-  conduit::Node actions;
+  int mpi_comm = ascent_opts.has_child("mpi_comm") ? ascent_opts["mpi_comm"].to_int() : 0;
```
Should the case were we don't have a comm be `-1`, to trigger the right error conditions?

cyrush reviewed (commented):
You are on the right track. We don't want to expose the internal method, but we want a similar set of logic here. 

The goal is to read the actions on rank 0, and then broadcast to others so all tasks have the same actions passed to `ascent.execute`.  You have most of the logic in place -- added suggestions for next steps. 


We don't have any tests that run replay wired up into our build system.

We should explore adding those, I can help you with the steps.

emily-howell committed:
Applying recommended changes
1) passing actions conduit node by reference rather than returning a node
2) Broadcasting actions to other ranks if they are loaded correctly
3) Checking if file exists only on rank 0
4) Changing default comms from 0 to -1
5) Adding a warning for if the file is not found

cyrush commented on src/utilities/replay/replay.cpp:
```diff
@@ -156,9 +156,7 @@ void trim(std::string &s)
-conduit::Node load_actions(const std::string &file_name, int mpi_comm_id){
-  conduit::Node actions;
-
+void load_actions(conduit::Node &actions, const std::string &file_name, int mpi_comm_id){
```
Unwritten rule: prefer out params last :-)
```suggestion
void load_actions(const std::string &file_name, int mpi_comm_id, conduit::Node &actions)
{
```

cyrush reviewed (approved):
Looks good!

emily-howell committed:
Moving out param to end of param list

Co-authored-by: Cyrus Harrison <cyrush@llnl.gov>

emily-howell committed:
Moving out param to end of param list #2
